### ISSUE TITLE
Unpacking a dict produces error unless all remaining parameters have compatible type or the dict is of Any

### ISSUE THREAD

**Issue by moskupols** [microsoft/pyright#5486]
**Describe the bug**
Unpacking a dict produces error unless all remaining parameters have compatible type or the dict is of Any
**Code or Screenshots**
```python
def foo(a: int = 42, b: str = 'default'):
    pass

foo(**{'a': 100})
```

**To Reproduce**
Run pyright of version 1.1.312 or newer on the snippet above. Here's the error I'm getting:
```
  wut.py:5:7 - error: Argument of type "int" cannot be assigned to parameter "b" of type "str" in function "foo"
    "int" is incompatible with "str" (
```

**Expected behavior**
No error is reported.

**VS Code extension or command-line**
I've tried pyright 1.1.312 through 1.1.317 with CPython3.11

**Additional context**
Note that if I replace the dict with `{'a': typing.cast(typing.Any, 100)}`, the error is gone.

The snippet above is somewhat unrealistic because I tried to produce a minimal example. The real example we have looks more like this:
```python
def foo(a: int, b: str, c: int = 42, d: str = 'unknowable default'):
    ...

def call_foo(a: int, b: str, condition: bool):
    foo(a, b, **({d: 'custom value'} if condition else {}))
```

**Comment by erictraut** [comment]
Yes, this is intended behavior. The behavior changed in pyright 1.1.312 to match mypy's behavior. I can defend both behaviors, but since mypy has set the precedent here, it's best for pyright to follow suit. When pyright and mypy differ in argument matching rules, it affects overload matching, and that makes it difficult for library and stub authors to target both mypy and pyright.
